### ISSUE TITLE
Disabling asserts skips checks on `halide_malloc()` results (at least in C++ backend)

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#7253]
What it says: `no_asserts` skips insertion of checks for these (at least in the C++ backend, haven't checked LLVM yet). I'm gonna argue that there is no situation where we should ever omit runtime checks for malloc failure.

**Comment by abadams** [comment]
I don't think we should be in the business of picking and choosing which asserts a user really wants, despite saying no asserts. No asserts should mean no asserts.

Maybe what you want is a NoBoundsChecking feature which disables bounds-related asserts and bounds querying?

**Comment by steven-johnson** [comment]
> I don't think we should be in the business of picking and choosing which asserts a user really wants, despite saying no asserts. No asserts should mean no asserts.

The problem here is that since this feature exists, and can save a small amount of runtime (1-3%), end users will enable it without realizing the danger. If the bulk of the cost is in bounds-checking, we should add a way to disable those.

**Comment by abadams** [comment]
I have some use-cases for which I really want no asserts of any kind, and I don't want to have to add a NoAssertsIReallyMeanIt feature.

**Comment by steven-johnson** [comment]
> I have some use-cases for which I really want no asserts of any kind, and I don't want to have to add a NoAssertsIReallyMeanIt feature.

Sure, I can see that, but maybe what we want is to make it more granular, so that instead of `no_asserts` we have `no_bounds_checks`, `no_allocation_checks`, etc, with the current `no_asserts` being the union of all of them, so that naive users won't reach for maximal-turn-off-all-safety via a single feature.

**Comment by steven-johnson** [comment]
(See also https://github.com/halide/Halide/issues/7255)
